### PR TITLE
Uniform sampling

### PR DESCRIPTION
**Pull request by kjs73**
Adds uniform sampling of points in n-dimensional spheres and cubes.
This is useful for some direct sampling applications.

### CODE REVIEW

File: source/mcpele/uniform_spherical_sampling.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+          m_dist(-radius, radius)
+    {}
+    void set_generator_seed(const size_t inp) { m_gen.seed(inp); }
+    virtual void displace(pele::Array<double>& coords, MC* mc)
+    {
+        do {
+            for (size_t i = 0; i < coords.size(); ++i) {
+                coords[i] = m_dist(m_gen);
+            }
+        } while (pele::dot(coords, coords) > m_radius2);
```
Comment (js850) [inline source/mcpele/uniform_spherical_sampling.h]: This type of rejection sampling will be exponentially slow for large dimensions when the volume of the sphere becomes minuscule compared to the volume of a cube.  I would guess that even for 3-dimensions it is very slower.  You can do this exactly, and without rejections, by first sampling a unit vector in n-dimensions, then sampling the radius from a power law distribution.  I can look up the exact details if you need them.  @smcantab and I used this a lot to do the nested sampling stuff.

File: examples/compute_pi/compute_pi.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+"""
+This example computes the numerical value of Pi by Monte Carlo.
```
Comment (js850) [inline examples/compute_pi/compute_pi.py]: cool!

File: source/mcpele/uniform_spherical_sampling.h (hunk 1 above)
Comment (kjs73) [inline source/mcpele/uniform_spherical_sampling.h]: Thanks for your feedback! To me it was not clear what was actually faster for what number of dimensions:
http://mathworld.wolfram.com/BallPointPicking.html
But I would be happy to do everything with the rejection free method if that is fast.

**Comment by js850** [comment]
cool, I never saw that particular method, I wonder if it's actually faster than the method I mentioned.  I guess the only difference is drawing from an exponential distribution vs a power law distribution.

yeah, 3 dimensions is a marginal case.  You would have to do benchmarking to be sure.  But probably the takestep routine is not going to be the bottleneck for most calculations, so I would use the one that generalizes well.

**Comment by kjs73** [comment]
Spherical sampling is now changed to the method described in the reference and, e.g. in Numerical Recipes, 3rd ed., p. 1129. 
Can this be merged?

File: source/mcpele/uniform_spherical_sampling.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    void set_generator_seed(const size_t inp) { m_gen.seed(inp); }
+    virtual void displace(pele::Array<double>& coords, MC* mc)
+    {
+        for (size_t i = 0; i < coords.size(); ++i) {
+            coords[i] = m_dist_normal(m_gen);
+        }
+        const double tmp = std::pow(pele::dot(coords, coords), -0.5) * m_radius * std::pow(m_dist_uniform(m_gen), 1 / static_cast<double>(coords.size()));
+        for (size_t i = 0; i < coords.size(); ++i) {
+            coords[i] *= tmp;
+        }
```
Comment (js850) [inline source/mcpele/uniform_spherical_sampling.h]: you can replace the loop with `coords *= tmp`

File: source/mcpele/uniform_spherical_sampling.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+          m_dist_normal(0, 1),
+          m_dist_uniform(0, 1)
+    {}
+    void set_generator_seed(const size_t inp) { m_gen.seed(inp); }
+    virtual void displace(pele::Array<double>& coords, MC* mc)
+    {
+        for (size_t i = 0; i < coords.size(); ++i) {
+            coords[i] = m_dist_normal(m_gen);
+        }
+        const double tmp = std::pow(pele::dot(coords, coords), -0.5) * m_radius * std::pow(m_dist_uniform(m_gen), 1 / static_cast<double>(coords.size()));
```
Comment (js850) [inline source/mcpele/uniform_spherical_sampling.h]: I find this line a bit hard to read.  Perhaps you could add a comment explaining it.  And maybe replace `std::pow(pele::dot(coords, coords), -0.5)` with `1.0 / norm(coords)`.

And replace `1 / static_cast<double>(coords.size())` with `1.0 / coords.size()`.  By explicitly writing 1.0 you ensure floating point division.

norm:  https://github.com/pele-python/pele/blob/master/source/pele/array.h#L464

**Comment by js850** [comment]
Hi Julian, it looks good. My comments were just suggestions, so i would say pull whenever you're ready.  (I didn't check whether it was correct, I'll assume it works as it's supposed to).

**Pull request merged**